pkg/broker/ingress/stats_reporter.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ingress
+
+import (
+	"context"
+	"log"
+	"strconv"
+	"time"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+	"go.uber.org/zap"
+	"knative.dev/eventing/pkg/logging"
+	"knative.dev/pkg/metrics"
+	"knative.dev/pkg/metrics/metricskey"
+)
+
+// stats_exporter is adapted from knative.dev/eventing/pkg/broker/ingress/stats_reporter.go
+// with the following changes:
+// - Metric descriptions are updated to match GCP broker specifics.
+// - Removed StatsReporter interface and directly use helper methods instead.
+
+var (
+	// dispatchTimeInMsecM records the time spent dispatching an event to
+	// a decouple queue, in milliseconds.
+	dispatchTimeInMsecM = stats.Float64(
+		"event_dispatch_latencies",
+		"The time spent dispatching an event to the decouple topic",
+		stats.UnitMilliseconds,
+	)
+
+	// Create the tag keys that will be used to add tags to our measurements.
+	// Tag keys must conform to the restrictions described in
+	// go.opencensus.io/tag/validate.go. Currently those restrictions are:
+	// - length between 1 and 255 inclusive
+	// - characters are printable US-ASCII
+	namespaceKey         = tag.MustNewKey(metricskey.LabelNamespaceName)
+	brokerKey            = tag.MustNewKey(metricskey.LabelBrokerName)
+	eventTypeKey         = tag.MustNewKey(metricskey.LabelEventType)
+	responseCodeKey      = tag.MustNewKey(metricskey.LabelResponseCode)
+	responseCodeClassKey = tag.MustNewKey(metricskey.LabelResponseCodeClass)
+	podKey               = tag.MustNewKey(metricskey.PodName)
+	containerKey         = tag.MustNewKey(metricskey.ContainerName)
+)
+
+type reportArgs struct {
+	namespace    string
+	broker       string
+	eventType    string
+	responseCode int
+}
+
+func init() {
+	register()
```

Comment:
Failure to register metrics should be fatal

Reply:
Is error better? I think we should still allow the application to run.

Reply:
This should not happen unless views are misconfigured. In this case it's better for the issue to be fatal so it is caught right away.

Reply:
SGTM.